### DISCUSSION TITLE
DAI 2024-2025 - Java IOs benchmarking - Alex Berberat

### DISCUSSION THREAD

**Discussion by AlexB-HEIG** [heig-vd-dai-course/heig-vd-dai-course#406] [Show and tell]
Link to repo : https://github.com/AlexB-HEIG/05-java-ios

# Observation
## Binary
|       | 1B     | 1KiB   | 1MiB    | 5MiB    |
|-------|--------|--------|---------|---------|
| Write | 163 ms | 160 ms | 1214 ms | 5156 ms |
| Read  | 156 ms | 179 ms | 477 ms  | 1529 ms |

## Buffered Binary
|       | 1B     | 1KiB   | 1MiB   | 5MiB   |
|-------|--------|--------|--------|--------|
| Write | 162 ms | 198 ms | 186 ms | 279 ms |
| Read  | 148 ms | 218 ms | 228 ms | 263 ms |

## Text
|       | 1B     | 1KiB   | 1MiB   | 5MiB   |
|-------|--------|--------|--------|--------|
| Write | 159 ms | 180 ms | 274 ms | 428 ms |
| Read  | 178 ms | 163 ms | 324 ms | 462 ms |

## Buffered Text
|       | 1B     | 1KiB   | 1MiB   | 5MiB   |
|-------|--------|--------|--------|--------|
| Write | 261 ms | 170 ms | 189 ms | 331 ms |
| Read  | 190 ms | 160 ms | 254 ms | 273 ms |

# Questions
## Which type of stream is the most efficient for each use case?
To write the buffered binary is the most efficient.  
To read the buffered binary and buffered text are the most efficient.  

## Why is it more efficient than the other types of streams?
When buffered less call are made thus the efficiency improvement.  

## What is the difference between binary data and text data?
Binary data stores the data directly as 0s and 1s.  
Text data will need to be encoded over multiple Bytes using different charsets.  

## What is a character encoding?
An indication of how the characters have been stored and how it needs to be interpreted so that it is readable and usable by a human.

## Why is this benchmark methodology important?
Because it allows us to see the importance of choosing the right type of data interaction.

**Comment by ludelafo** [comment]
Thank you!

You can always find the solution in the solution repository if you ever need to compare your work with the official one here: https://github.com/heig-vd-dai-course/heig-vd-dai-course-solutions.

Do not hesitate to ask questions if something is not clear, either in this discussion or in the Q&A category of the GitHub Discussions here: https://github.com/orgs/heig-vd-dai-course/discussions/categories/q-a.
